### PR TITLE
Enhancements for Tutorial Notebooks and Comprehensive Support for AnnData

### PR DESCRIPTION
**Pull request by katosh**
This Pull Request (PR) consolidates all the proposed modifications aimed at improving the tutorial notebooks we are preparing for the Mellon publication. The intention is to foster constructive discussions and considerations about these changes before submitting the PR to the main repository, [dpeerlab/Palantir](https://github.com/dpeerlab/Palantir).

A major element of this PR is the introduction of enhanced support for AnnData objects. This upgrade standardizes and automates our established best practices when working with AnnData and Palantir, and ensures that these improvements don't compromise the existing functionalities.

The principal commit incorporating these extensive changes is 3607b44b5d0583be843b833fba85267c934a1c00.

### CODE REVIEW

**Comment by ManuSetty** [comment]
Added some comments in the commit linked above - happy to discuss further!

**Comment by katosh** [comment]
Rplacing PhenoGraph with leiden in 4f935642cbe19d618b3ea599041fb30e0085de63 changes the results of the tutorial notebook since leiden finds 10 clusters with its default resolution:

image

**Comment by katosh** [comment]
`palantir.utils.run_magic_imputation` now takes any input:
image

**Comment by ManuSetty** [comment]
> Rplacing PhenoGraph with leiden in [4f93564](https://github.com/settylab/Palantir/commit/4f935642cbe19d618b3ea599041fb30e0085de63) changes the results of the tutorial notebook since leiden finds 10 clusters with its default resolution:
> 
> image

These clusters do look quite good to me.

File: src/palantir/presults.py [hunk 1]
```diff
@@ -195,45 +253,79 @@ def _gam_fit_predict_rpy2(x, y, weights=None, pred_x=None):
...
     return y_pred, stds
 
 
-def cluster_gene_trends(trends, k=150, n_jobs=-1):
-    """Function to cluster gene trends
-    :param trends: Matrix of gene expression trends
-    :param k: K for nearest neighbor construction
-    :param n_jobs: Number of jobs for parallel processing
-    :return: Clustering of gene trends
+def cluster_gene_trends(
```
Comment (ManuSetty) [inline src/palantir/presults.py:387]: Can we please add an optional `var` key if Anndata is specified to indicate which genes need to be clustered? I dont think we would want to cluster all genes at any point.

File: src/palantir/utils.py [hunk 2]
```diff
@@ -113,81 +196,417 @@ def run_diffusion_maps(data_df, n_components=10, knn=30, alpha=0, seed=None):
...
+        data.obsp[kernel_key] = res["kernel"]
+        data.obsm[eigvec_key] = res["EigenVectors"].values
+        data.uns[eigval_key] = res["EigenValues"].values
+
     return res
 
 
-def run_magic_imputation(data, dm_res, n_steps=3):
-    """Run MAGIC imputation
+def _dot_herlper_func(x, y):
```
Comment (ManuSetty) [inline src/palantir/utils.py]: Small typo here. `herlper` -> `helper`

**Comment by katosh** [comment]
The last commit 68af21fb8c777a4708c3a8b3320532e1951e7b9e depends on a version of mellon that has not been released yet. Please install it with
```
pip install --no-deps 'git+https://github.com/settylab/Palantir@mellon'
pip install 'git+https://github.com/settylab/Mellon'
```

**Comment by katosh** [comment]
- Gene trends can now be compared again.
 - I also finished implementing the AnnData centered workflow throughout all functions used in the [tutorial notebook](https://github.com/settylab/Palantir/blob/mellon/notebooks/Palantir_sample_notebook.ipynb).
 - I added type hints and extended the documentation of all touched functions.
 - While almost everything is backwards compatible with the previous workflow of passing dataframes and dictionaries, I updated the tutorial to reflect the new AnnData centered workflow for ease of use.
 - Since we removed `run_tsne` (since scanpy covers this) I also removed `palantir.utils.determine_cell_clusters` and `palantir.plot.plot_cell_clusters` for the same reason

For testing install with
```
pip install --no-deps 'git+https://github.com/settylab/Palantir@mellon'
pip install 'git+https://github.com/settylab/Mellon'
```

@ManuSetty This is all from my side for now. Do you have further suggestions? When we release Mellon v1.3.0 (a new dependency) then we can make the PR to the official repo if you like.

File: src/palantir/presults.py (hunk 1 above)
Comment (katosh) [reply src/palantir/presults.py:387]: With bb579cc72e3c6594530624ea3ca5bed9c07725bb the function now takes a gene list in the `genes` argument.

**Comment by ManuSetty** [comment]
I just noticed in cell 2 of the notebooks that the anndata load says `pbmc_10x_rna_processed.h5ad` - this should be `marrow_counts...`

**Comment by katosh** [comment]
You are right! Fixed it with 99f138d9c87b7036b2a306d50d7802a3802327b9

**Comment by katosh** [comment]
Mellon v1.3.0 is now available in pypi: https://pypi.org/project/mellon/ so the new dependency is fulfillable.

The last thing before merge should be the implementation of local gene variability.

**Pull request merged**